wrappers/kmip/kmip.go

```diff
@@ -0,0 +1,537 @@
+// Copyright (c) 2025 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+// Initially inspired and adapted from ceph's internal kmip package
+// https://github.com/ceph/ceph-csi/blob/devel/internal/kms/kmip.go
+
+package kmip
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+	"time"
+
+	uuid "github.com/hashicorp/go-uuid"
+	wrapping "github.com/openbao/go-kms-wrapping/v2"
+	"github.com/ovh/kmip-go"
+	"github.com/ovh/kmip-go/kmipclient"
+	"github.com/ovh/kmip-go/ttlv"
+)
+
+// These constants contain the accepted env vars; the Vault one is for backwards compat
+const (
+	EnvKmipWrapperKeyId   = "KMIP_WRAPPER_KEY_ID"
+	EnvVaultKmipSealKeyId = "VAULT_KMIP_SEAL_KEY_ID"
+
+	EnvKmipEndpoint     = "KMIP_ENDPOINT"
+	EnvKmipCaCert       = "KMIP_CA_CERT"
+	EnvKmipClientCert   = "KMIP_CLIENT_CERT"
+	EnvKmipClientKey    = "KMIP_CLIENT_KEY"
+	EnvKmipServerName   = "KMIP_SERVER_NAME"
+	EnvKmipTimeout      = "KMIP_TIMEOUT"
+	EnvKmipEncryptAlg   = "KMIP_ENCRYPT_ALG"
+	EnvKmipTls12Ciphers = "KMIP_TLS12_CIPHERS"
+)
+
+var (
+	// Preset kmip cryptographic parameters that can be selected from config
+	cryptoParamsPreset = map[string]kmip.CryptographicParameters{
+		"AES_GCM": {
+			CryptographicAlgorithm: kmip.CryptographicAlgorithmAES,
+			BlockCipherMode:        kmip.BlockCipherModeGCM,
+			TagLength:              16,
+			IVLength:               12,
+		},
+		"RSA_OAEP_SHA256": {
+			CryptographicAlgorithm:        kmip.CryptographicAlgorithmRSA,
+			PaddingMethod:                 kmip.PaddingMethodOAEP,
+			HashingAlgorithm:              kmip.HashingAlgorithmSHA_256,
+			MaskGenerator:                 kmip.MaskGeneratorMGF1,
+			MaskGeneratorHashingAlgorithm: kmip.HashingAlgorithmSHA_256,
+		},
+		"RSA_OAEP_SHA384": {
+			CryptographicAlgorithm:        kmip.CryptographicAlgorithmRSA,
+			PaddingMethod:                 kmip.PaddingMethodOAEP,
+			HashingAlgorithm:              kmip.HashingAlgorithmSHA_384,
+			MaskGenerator:                 kmip.MaskGeneratorMGF1,
+			MaskGeneratorHashingAlgorithm: kmip.HashingAlgorithmSHA_384,
+		},
+		"RSA_OAEP_SHA512": {
+			CryptographicAlgorithm:        kmip.CryptographicAlgorithmRSA,
+			PaddingMethod:                 kmip.PaddingMethodOAEP,
+			HashingAlgorithm:              kmip.HashingAlgorithmSHA_512,
+			MaskGenerator:                 kmip.MaskGeneratorMGF1,
+			MaskGeneratorHashingAlgorithm: kmip.HashingAlgorithmSHA_512,
+		},
+	}
+
+	// Map between crypto param name, and mechanism ID. The ID is saved with the wrapped key
+	// so we can use the same for decryption. This allows changing the cryptographic parameter for the next encryption.
```

Comment:
I'm confused by this comment; couldn't we range through `cryptoParamsPreset`'s keys?

Reply:
ranging through a map in golang does not guaranty the order. But I could change the code to store cryptoParamsPresets in an array with their keys, and lookup by ranging through it
```go
type CryptoParamPreset struct {
    name string
    mechanismId int
    params kmip.CryptographicParameter
}

var cryptoParamsPreset = []CryptoParamPreset{ .... }

func getPresetByName(name string) (CryptoParamPreset, bool) {}
func getPresetById(id int) (CryptoParamPreset, bool) {}
```

I don't have any preference, do you ?

Reply:
Ohh, I see. I was confused by:

> This allows changing the cryptographic parameter for the next encryption.

One of the shortcomings of go-kms-wrapping is that each instance of the wrapper has a fixed config. Changing the encryption algorithm isn't necessarily possible mid-use. You'd spin up a new wrapper instance entirely if you wanted to do that.

I was thinking you were trying to support on-the-fly algorithm switches.

But, what you're saying is someone might have already encrypted a key and now want to decrypt it under the existing config, so we need a canonical ordering of KMIP mechanism name to integer identifier.

I think this is fine.

Reply:
Yeah the comment is maybe a bit misleading